### ISSUE TITLE
Missing User ID for NDK reports

### ISSUE THREAD

**Issue by sgong-pdftron** [microsoft/appcenter-sdk-android#1497]
### **Description**

Please describe the issue you are facing using the SDK.

With "AppCenter.setUserId(uuid)", Java crash reports are able to show User ID in the App Center console.
However, NDK crash reports show "N/A" for User ID.

![image](https://user-images.githubusercontent.com/20527369/103708059-a516f080-4f64-11eb-8e52-fb1e2e19602c.png)

![image](https://user-images.githubusercontent.com/20527369/103708040-992b2e80-4f64-11eb-8861-7eca9009dd29.png)

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Setup SDK as described here: https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/android
2. Setup Native crash as described here: https://docs.microsoft.com/en-us/appcenter/sdk/crashes/android#reporting-ndk-crashes

### **Details**

1. Which SDK version are you using?
    - e.g. 4.1.0
2. Which OS version did you experience the issue on?
    - e.g. Android 7.1.1
3. What device version did you see this error on?  Were you using an emulator or a physical device?
    - e.g. Nexus 9 physical device
4. What third party libraries are you using?
    - Firebase

5. Please enable verbose logging for your app using `AppCenter.setLogLevel(Log.VERBOSE)` before your call to `AppCenter.start(...)` and include the logs here:

[ndk-crash-log.zip](https://github.com/microsoft/appcenter-sdk-android/files/5773014/ndk-crash-log.zip)

**Comment by AnastasiaKubova** [comment]
Hi! Thanks for getting in touch with us! I checked this behavior on the latest version of App Center SDK and it works right for me. Could you please share a code snippet with use our SDK and full logs before and after crash?  Do you have similar behavior in the previous versions?

**Comment by sgong-pdftron** [comment]
@AnastasiaKubova 4.1.0 is the first App Center SDK I used so can't compare with other versions.

Code snippet:

            AppCenter.start(((Activity) context).getApplication(), getHockeyappId(), Analytics.class, Crashes.class);
            // update uuid
            String uuid = AppCenter.getInstallId().get().toString();
            AppCenter.setUserId(uuid);

            Crashes.getMinidumpDirectory().thenAccept(path -> {
                // Path is null when Crashes is disabled.
                if (path != null) {
                    AHBreakpadHelper.setUpBreakpad(path);
                }
            });

I've attached full log in my initial post, does that help?
Just to make sure we are on the same page, I'm only not seeing the User ID for NDK crashes, Java crashes show User ID fine.

**Comment by AnastasiaKubova** [comment]
Based on your attached logs I don't see that the `userId` value was set and sent with the crash report. Additional logs will helpful in investigating your issue. Could you please share the verbose logs before and after the crash?

**Comment by sgong-pdftron** [comment]
[Archive.zip](https://github.com/microsoft/appcenter-sdk-android/files/5798396/Archive.zip)

@AnastasiaKubova Here are 2 logs, one for session that produced crash, one for session after that. I added a log where I called:

            AppCenter.setUserId(uuid);
            Log.d("AppCenter", "AppCenter.setUserId: " + uuid);

Could you check? Thanks.

**Comment by AnastasiaKubova** [comment]
Thanks for share additional info. Could you please clarify how you crash your app?

**Comment by sgong-pdftron** [comment]
@AnastasiaKubova It was a test crash with division by zero in c++

**Comment by russelarms** [comment]
Hi @sgong-pdftron, it's a known limitation of App Center Android SDK. User Id for NDK crashes is applied only after restart:
https://github.com/microsoft/appcenter-sdk-android/blob/develop/sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java#L782:L787

Normally you have to set **userId** after `AppCenter.start()` which sometimes can lead to a race condition when an NDK crash is processed without **userId**.

To work around this you can set the **userId** explicitly before `AppCenter.start()`:
```
UserIdContext.getInstance().setUserId("user id");
AppCenter.start(...);
```

Since we have a _TODO_ for this in our code I'll probably convert this issue into a small feature request.

**Comment by sgong-pdftron** [comment]
@russelarms the problem is for me I am using `AppCenter.getInstallId` to set as the ID, however doesn't seem to be available before start. Do you have any suggestions?

**Comment by AnastasiaKubova** [comment]
Hi there! This fix has been released in the [4.1.1](https://github.com/microsoft/appcenter-sdk-android/releases/tag/4.1.1) version of App Center SDK. Please pay attention that due to the termination of `jCenter` support all our assemblies were moved to the `Maven Central` repository.

See more about migration to `Maven Central` [here](https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/android#3-add-the-app-center-sdk-modules).
